### PR TITLE
fix(api-machinery): Fixes patching for unstructured objects

### LINKED ISSUE #3382: Helm 2.8.0 with CRDs: unable to find api field in struct Unstructured for the json field "metadata"
Hi, 
We're seeing this a lot since updating to 2.8.0.  This only happens on subsequent upgrades to existing charts.

```
❯ helm upgrade --reset-values --debug --wait --install --namespace default --version 0.1.379 at-sauron-web-mobile . -f values.yaml -f dev.yaml
[debug] Created tunnel using local port: '56621'

[debug] SERVER: "127.0.0.1:56621"

Error: UPGRADE FAILED: failed to create patch: unable to find api field in struct Unstructured for the json field "metadata"
```

This particular chart is upgrading the following components:
  - configmap
  - service
  - deployment 
  - ingress

If i take the `--dry-run` output and put it into a file and do `kubectl apply -f` myself, it works fine

### PR DESCRIPTION
**Pull request by mattfarina**
CRDs and other objects seen as unstructured cannot use strategic
merge patching. It has never been supported on CRDs. Previously,
cases like unstructured objects could have caused an unregistered
error. This is no longer the case.

This change explicitly looks for unstructured objects and handles
those using json merge patching.

/cc @bacongobbler @Stono @adamreese @technosophos @liggitt 

Closes #3382

### CODE REVIEW

**Comment by Stono** [comment]
Sweet, seems to solve it:
```
/Users/karl.stoney/git/other/bad-chart-2.8-1.9 git/master *
❯ helm upgrade --reset-values --install --namespace default foo .
Release "foo" has been upgraded. Happy Helming!
LAST DEPLOYED: Mon Feb  5 21:06:09 2018
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1alpha2/RouteRule
NAME                 AGE
some-service-policy  33s
```

`k get routerule some-service-policy` === updated values.  Great stuff :-)

**Comment by Stono** [comment]
And for bonus confidence this is me running 17 CRDs with modified metadata through it:

```
[kube] 2018/02/05 21:11:01 beginning wait for 17 resources with timeout of 5m0s
[tiller] 2018/02/05 21:11:03 executing 0 post-upgrade hooks for istio-crd
[tiller] 2018/02/05 21:11:03 hooks complete for post-upgrade istio-crd
[storage] 2018/02/05 21:11:03 updating release "istio-crd.v4"
[tiller] 2018/02/05 21:11:03 updating status for updated release for istio-crd
[storage] 2018/02/05 21:11:03 updating release "istio-crd.v5"
[storage] 2018/02/05 21:11:03 getting last revision of "istio-crd"
[storage] 2018/02/05 21:11:03 getting release history for "istio-crd"
[kube] 2018/02/05 21:11:03 Doing get for attributemanifest: "istioproxy"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/attributemanifest/istioproxy
[kube] 2018/02/05 21:11:03 Doing get for attributemanifest: "kubernetes"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/attributemanifest/kubernetes
[kube] 2018/02/05 21:11:03 Doing get for kubernetes: "attributes"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/kubernetes/attributes
[kube] 2018/02/05 21:11:03 Doing get for kubernetesenv: "handler"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/kubernetesenv/handler
[kube] 2018/02/05 21:11:03 Doing get for logentry: "accesslog"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/logentry/accesslog
[kube] 2018/02/05 21:11:03 Doing get for metric: "requestcount"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/metric/requestcount
[kube] 2018/02/05 21:11:03 Doing get for metric: "requestduration"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/metric/requestduration
[kube] 2018/02/05 21:11:03 Doing get for metric: "requestsize"
[kube] 2018/02/05 21:11:03 get relation pod of object: istio-system/metric/requestsize
[kube] 2018/02/05 21:11:03 Doing get for metric: "responsesize"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/metric/responsesize
[kube] 2018/02/05 21:11:04 Doing get for metric: "tcpbytereceived"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/metric/tcpbytereceived
[kube] 2018/02/05 21:11:04 Doing get for metric: "tcpbytesent"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/metric/tcpbytesent
[kube] 2018/02/05 21:11:04 Doing get for prometheus: "handler"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/prometheus/handler
[kube] 2018/02/05 21:11:04 Doing get for rule: "kubeattrgenrulerule"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/rule/kubeattrgenrulerule
[kube] 2018/02/05 21:11:04 Doing get for rule: "promhttp"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/rule/promhttp
[kube] 2018/02/05 21:11:04 Doing get for rule: "promtcp"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/rule/promtcp
[kube] 2018/02/05 21:11:04 Doing get for rule: "stdio"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/rule/stdio
[kube] 2018/02/05 21:11:04 Doing get for stdio: "handler"
[kube] 2018/02/05 21:11:04 get relation pod of object: istio-system/stdio/handler
```

File: pkg/kube/client.go [hunk 1]
```diff
@@ -399,14 +399,27 @@ func createPatch(mapping *meta.RESTMapping, target, current runtime.Object) ([]b
...
+	// returned from ConvertToVersion. Anything that's unstructured should
+	// use the jsonpatch.CreateMergePatch. Strategic Merge Patch is not supported
+	// on objects like CRDs.
+	_, isUnstructured := versionedObject.(runtime.Unstructured)
+
 	if apiequality.Semantic.DeepEqual(oldData, newData) {
+		// Unstructured types, such as CRDs, are not supported by Strategic
+		// Merge Patches
+		if isUnstructured {
+			return nil, types.MergePatchType, nil
```
Comment (adamreese) [inline pkg/kube/client.go]: The line below is already returning a NOP.  This can be deleted.

**Review by adamreese — APPROVED** [review]
Validated patch works against https://github.com/mattfarina/bad-chart-2.8-1.9

File: pkg/kube/client.go (hunk 1 above)
Comment (mattfarina) [reply pkg/kube/client.go]: I'll work on an update. Thanks.

**Comment by mattfarina** [comment]
@adamreese I updated it and put a comment there so future folks don't need to figure out what you showed me.

**Review by SlickNik — APPROVED** [review]
Thanks for fixing this @mattfarina! 👍

**Pull request merged**